File: frontend/src/types/game.ts
export type AnswerLetter = 'A' | 'B' | 'C' | 'D'

export type AnswerFailureReason = 'INCORRECT' | 'TIMEOUT'

export type GameStatus =
  | 'LOBBY'
  | 'QUESTION'
  | 'BUZZ_OPEN'
  | 'BUZZ_LOCKED'
  | 'ANSWER_RESULT'
  | 'SCOREBOARD'
  | 'FINISHED'

export interface QuestionOption {
  letter: AnswerLetter
  text: string
}

export interface Question {
  id: number
  text: string
  options: QuestionOption[]
  correct_answer: AnswerLetter
  explanation: string
  order: number
  active?: boolean
}

export interface GamePlayer {
  id: number
  player_number: number
  name: string
  color: string
  score: number
  active: boolean
}

export interface BuzzPlayer {
  id: number
  player_number: number
  name: string
}

export interface AnswerResult {
  player_id: number
  player_name: string
  selected_answer: AnswerLetter | ''
  is_correct: boolean
  points_awarded: number
  correct_answer?: AnswerLetter | null
  correct_answer_text?: string | null
  explanation?: string | null
  reason?: AnswerFailureReason | null
}

export interface GameDetail {
  id: number
  created_at: string
  status: GameStatus
  current_question_index: number
  finished: boolean
  total_questions: number
  players: GamePlayer[]
  current_question: Question | null
  buzz_player: BuzzPlayer | null
  buzz_deadline: string | null
  failed_player_ids: number[]
  last_answer_result: AnswerResult | null
  reveal_correct_answer: boolean
}

export interface CreatePlayerInput {
  number: number
  name: string
}

export interface DisplayPlayer {
  id: number
  name: string
  color: string
  score: number
  avatarInitial: string
  playerNumber: number
}

export interface DisplayQuestion {
  number: number
  total: number
  text: string
  answers: {
    letter: AnswerLetter
    text: string
    color: string
  }[]
  correctLetter: AnswerLetter
  correctText: string
  explanation: string
}

export type WebSocketAction =
  | { action: 'lock_buzzers' }
  | { action: 'reset_buzzers' }
  | { action: 'register_buzz'; player_number: number }
  | { action: 'submit_answer'; answer: AnswerLetter }
  | { action: 'next_question' }
  | { action: 'finish_game' }

export type WebSocketEventType =
  | 'game_state'
  | 'buzz_registered'
  | 'answer_result'
  | 'question_changed'
  | 'game_finished'
  | 'error'

export interface WebSocketMessage {
  type: WebSocketEventType
  payload: unknown
}

File: backend/app/schemas/game.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.models.game_status import GameStatus
from app.schemas.question import QuestionPublicResponse, QuestionResponse

AnswerFailureReason = Literal["INCORRECT", "TIMEOUT"]


class CreatePlayerRequest(BaseModel):
    number: int = Field(ge=1, le=99)
    name: str = Field(min_length=1, max_length=20)

    @field_validator("name")
    @classmethod
    def strip_name(cls, name: str) -> str:
        return name.strip()


class CreateGameRequest(BaseModel):
    question_set_id: int = Field(ge=1)
    players: list[CreatePlayerRequest] = Field(min_length=1, max_length=99)

    @field_validator("players")
    @classmethod
    def validate_unique_player_numbers(cls, players: list[CreatePlayerRequest]) -> list[CreatePlayerRequest]:
        numbers = [player.number for player in players]
        if len(numbers) != len(set(numbers)):
            raise ValueError("Los números de jugador no pueden repetirse")
        return players


class PlayerResponse(BaseModel):
    id: int
    player_number: int
    name: str
    color: str
    score: int
    active: bool


class BuzzPlayerResponse(BaseModel):
    id: int
    player_number: int
    name: str


class AnswerResultResponse(BaseModel):
    player_id: int
    player_name: str
    selected_answer: str
    is_correct: bool
    points_awarded: int
    correct_answer: str | None = None
    correct_answer_text: str | None = None
    explanation: str | None = None
    reason: AnswerFailureReason | None = None


class GameResponse(BaseModel):
    id: int
    created_at: str
    status: GameStatus
    current_question_index: int
    finished: bool
    total_questions: int
    players: list[PlayerResponse]
    current_question: QuestionPublicResponse | None = None
    buzz_player: BuzzPlayerResponse | None = None
    buzz_deadline: str | None = None
    failed_player_ids: list[int] = Field(default_factory=list)
    last_answer_result: AnswerResultResponse | None = None
    reveal_correct_answer: bool = False


class GameDetailResponse(GameResponse):
    current_question: QuestionResponse | None = None
